hazelcast/src/main/java/com/hazelcast/sql/impl/expression/CastExpression.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression;
+
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.converter.Converter;
+
+/**
+ * Implements evaluation of SQL CAST operator.
+ */
+public final class CastExpression<T> extends UniExpressionWithType<T> implements IdentifiedDataSerializable {
+
+    public CastExpression() {
+        // No-op.
+    }
+
+    private CastExpression(Expression<?> operand, QueryDataType resultType) {
+        super(operand, resultType);
+    }
+
+    public static CastExpression<?> create(Expression<?> operand, QueryDataType resultType) {
+        return new CastExpression<>(operand, resultType);
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_CAST;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T eval(Row row, ExpressionEvalContext context) {
+        Object value = operand.eval(row, context);
+
+        if (value == null) {
+            return null;
+        }
+
+        Converter fromConverter = operand.getType().getConverter();
```

Comment:
I am not sure whether this is worth, but - we rely on casts heavily now, therefore the performance is important. In principle, we may avoid calling converter at all if the object is already of expected type:
```
if (resultType.getConverter().getNormalizedValueClass() == value.getClass()) {
    return value;
}
```

Reply:
Casts should be inserted only if they are really needed, so there is no point in performing additional checks.

Reply:
How about a parameter placeholder, that is coerced to `BIGINT`, while the parameter itself is already `Long`?

Reply:
Parameters should never be coerced through cast insertion, they just get some type assigned during validation, there should be no casts produced. If a parameter is casted explicitly by user, the parameter receives a type derived from the cast: `cast(? as int)`, the parameter receives `int` type, the cast itself is optimized away. While preparing parameter arguments for a query, all the passed arguments are converted to their expected types once.